src/Shell.zig

```diff
@@ -0,0 +1,303 @@
+                    fmt_ansi = fmt_ansi ++ fmt[pos_start..pos] ++ @field(ansi, field_name);
+                    pos += tag.len;
+                    pos_start = pos;
+                    continue :next_pos;
+                }
+            }
+        }
+        pos += 1;
+    }
+    fmt_ansi = fmt_ansi ++ fmt[pos_start..pos] ++ "\n";
+
+    std.debug.print(fmt_ansi, args);
+}
+
+/// Checks if the path exists and is a directory.
+pub fn dir_exists(shell: *Shell, path: []const u8) !bool {
+    _ = shell;
+
+    const stat = std.fs.cwd().statFile(path) catch |err| switch (err) {
+        error.FileNotFound => return false,
+        error.IsDir => return true,
+        else => return err,
+    };
+
+    return stat.kind == .Directory;
+}
+
+/// Analogue of the `find` utility, returns a set of paths matching filtering criteria.
+///
+/// Returned slice is stored in `Shell.arena`.
+pub fn find(shell: *Shell, options: struct {
+    where: []const []const u8,
+    ends_with: []const u8,
+}) ![]const []const u8 {
+    var result = std.ArrayList([]const u8).init(shell.arena.allocator());
+
+    const cwd = std.fs.cwd();
+    for (options.where) |base_path| {
+        var base_dir = try cwd.openDir(base_path, .{ .iterate = true });
+        defer base_dir.close();
+
+        var walker = try base_dir.walk(shell.gpa);
+        defer walker.deinit();
+
+        while (try walker.next()) |entry| {
+            if (std.mem.endsWith(u8, entry.path, options.ends_with)) {
+                const full_path =
+                    try std.fs.path.join(shell.arena.allocator(), &.{ base_path, entry.path });
+                try result.append(full_path);
+            }
+        }
+    }
+
+    return result.items;
+}
+
+/// Runs the given command with inherited stdout and stderr.
+/// Returns an error if exit status is non-zero.
+///
+/// Supports interpolation using the following syntax:
+///
+/// ```
+/// shell.exec("git branch {op} {branches}", .{
+///     op = "-D",
+///     branches = &.{"main", "feature"},
+/// })
+/// ```
+pub fn exec(shell: Shell, comptime cmd: []const u8, args: anytype) !void {
+    var argv = Argv.init(shell.gpa);
+    defer argv.deinit();
+
+    try format_argv(&argv, cmd, args);
+
+    const child = try std.ChildProcess.init(argv.items, shell.gpa);
+    defer child.deinit();
+
+    child.stdin_behavior = .Ignore;
+    child.stdout_behavior = .Inherit;
+    child.stderr_behavior = .Inherit;
+
+    const term = try child.spawnAndWait();
+
+    switch (term) {
+        .Exited => |code| if (code != 0) return error.NonZeroExitStatus,
+        else => return error.CommandFailed,
+    }
+}
+
+/// Returns `true` if the command executed successfully with a zero exit code and an empty stderr.
+///
+/// The intended use-case is sanity-checking that an executable is present, by running
+/// `my-tool --version`.
+pub fn exec_status_ok(shell: *Shell, comptime cmd: []const u8, args: anytype) !bool {
```

Comment:
Seems very specific - and some utils might output on `stderr`... We could just use `exec` at the caller, since it seems we return `false` for any `exec` error here? Or, we could make this just wrap `exec` above and return `true` on void, `false` on `NonZeroExitStatus` and pass through the error otherwise.